File: modal-js/test/environment.test.ts
import { expect, test, vi } from "vitest";
import { ModalClient } from "../src/client";
import { MockGrpcClient } from "../test-support/grpc_mock";

function createTestJWT(exp: number): string {
  const header = "x";
  const payload = Buffer.from(JSON.stringify({ exp })).toString("base64");
  const signature = "x";
  return `${header}.${payload}.${signature}`;
}

function createMockEnvironmentClient(): {
  mockClient: ModalClient;
  mock: MockGrpcClient;
  getCallCount: () => number;
} {
  let callCount = 0;
  const mock = new MockGrpcClient();

  mock.handleUnary("/AuthTokenGet", () => {
    return { token: createTestJWT(9999999999) };
  });

  const originalHandleUnary = mock.handleUnary.bind(mock);
  mock.handleUnary = (rpcName: string, handler: any) => {
    if (rpcName === "/EnvironmentGetOrCreate") {
      originalHandleUnary(rpcName, (req: any) => {
        callCount++;
        return handler(req);
      });
    } else {
      originalHandleUnary(rpcName, handler);
    }
  };

  const mockClient = new ModalClient({
    cpClient: mock as any,
    tokenId: "test-token-id",
    tokenSecret: "test-token-secret",
  });

  return {
    mockClient,
    mock,
    getCallCount: () => callCount,
  };
}

test("GetEnvironmentCached", async () => {
  // Temporarily unset the env var so we fetch from server
  vi.stubEnv("MODAL_IMAGE_BUILDER_VERSION", undefined);

  const { mockClient, mock, getCallCount } = createMockEnvironmentClient();

  mock.handleUnary("/EnvironmentGetOrCreate", () => {
    return {
      environmentId: "en-test123",
      metadata: {
        name: "main",
        settings: {
          imageBuilderVersion: "2024.10",
          webhookSuffix: "modal.run",
        },
      },
    };
  });

  const version1 = await mockClient.getImageBuilderVersion();
  expect(version1).toBe("2024.10");
  expect(getCallCount()).toBe(1);

  const version2 = await mockClient.getImageBuilderVersion();
  expect(version2).toBe("2024.10");
  expect(getCallCount()).toBe(1); // got "" from cache

  const { mockClient: mockClientDev, mock: mockDev } = createMockEnvironmentClient();

  mockDev.handleUnary("/EnvironmentGetOrCreate", () => {
    return {
      environmentId: "en-dev",
      metadata: {
        name: "dev",
        settings: {
          imageBuilderVersion: "2025.06",
          webhookSuffix: "",
        },
      },
    };
  });

  // Create a new client with "dev" environment
  const newMockClientDev = new ModalClient({
    cpClient: mockDev as any,
    tokenId: "test-token-id",
    tokenSecret: "test-token-secret",
    environment: "dev",
  });

  const versionDev = await newMockClientDev.getImageBuilderVersion();
  expect(versionDev).toBe("2025.06");

  mockClient.close();
  mockClientDev.close();
  newMockClientDev.close();

  vi.unstubAllEnvs();
});

test("ImageBuilderVersion_LocalConfigHasPrecedence", async () => {
  const { mockClient, mock, getCallCount } = createMockEnvironmentClient();

  mock.handleUnary("/EnvironmentGetOrCreate", () => {
    return {
      environmentId: "en-test",
      metadata: {
        name: "",
        settings: {
          imageBuilderVersion: "2024.10",
          webhookSuffix: "",
        },
      },
    };
  });

  const mockClientWithConfig = new ModalClient({
    cpClient: mock as any,
    tokenId: "test-token-id",
    tokenSecret: "test-token-secret",
  });

  mockClientWithConfig.profile.imageBuilderVersion = "2024.04";

  const version = await mockClientWithConfig.getImageBuilderVersion();
  expect(version).toBe("2024.04");
  expect(getCallCount()).toBe(0); // should not fetch from server

  mockClient.close();
  mockClientWithConfig.close();
});

File: modal-go/environment_test.go
package modal

import (
	"context"
	"sync"
	"sync/atomic"
	"testing"

	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
	"github.com/onsi/gomega"
	"google.golang.org/grpc"
)

// mockAuthTokenResponse returns a mock AuthTokenGetResponse with a JWT that has a far-future
// expiration time. The JWT structure is: mock header + base64({"exp":9999999999}) + mock signature.
// This prevents warnings from AuthTokenManager.FetchToken() which expects an "exp" field.
func mockAuthTokenResponse() (*pb.AuthTokenGetResponse, error) {
	const mockJWT = "x.eyJleHAiOjk5OTk5OTk5OTl9.x"
	return pb.AuthTokenGetResponse_builder{Token: mockJWT}.Build(), nil
}

type mockEnvironmentClient struct {
	pb.ModalClientClient
	callCount atomic.Int64
	envs      sync.Map
}

func (m *mockEnvironmentClient) EnvironmentGetOrCreate(ctx context.Context, req *pb.EnvironmentGetOrCreateRequest, opts ...grpc.CallOption) (*pb.EnvironmentGetOrCreateResponse, error) {
	m.callCount.Add(1)
	if resp, ok := m.envs.Load(req.GetDeploymentName()); ok {
		return resp.(*pb.EnvironmentGetOrCreateResponse), nil
	}
	return pb.EnvironmentGetOrCreateResponse_builder{
		EnvironmentId: "en-default",
		Metadata: pb.EnvironmentMetadata_builder{
			Name: "",
			Settings: pb.EnvironmentSettings_builder{
				ImageBuilderVersion: "2024.10",
				WebhookSuffix:       "",
			}.Build(),
		}.Build(),
	}.Build(), nil
}

func (m *mockEnvironmentClient) AuthTokenGet(ctx context.Context, req *pb.AuthTokenGetRequest, opts ...grpc.CallOption) (*pb.AuthTokenGetResponse, error) {
	return mockAuthTokenResponse()
}

func (m *mockEnvironmentClient) getCallCount() int {
	return int(m.callCount.Load())
}

func TestGetEnvironmentCached(t *testing.T) {
	g := gomega.NewWithT(t)
	ctx := context.Background()

	t.Setenv("MODAL_IMAGE_BUILDER_VERSION", "")

	mockClient := &mockEnvironmentClient{}
	mockClient.envs.Store("", pb.EnvironmentGetOrCreateResponse_builder{
		EnvironmentId: "en-test123",
		Metadata: pb.EnvironmentMetadata_builder{
			Name: "main",
			Settings: pb.EnvironmentSettings_builder{
				ImageBuilderVersion: "2024.10",
				WebhookSuffix:       "modal.run",
			}.Build(),
		}.Build(),
	}.Build())
	mockClient.envs.Store("dev", pb.EnvironmentGetOrCreateResponse_builder{
		EnvironmentId: "en-dev",
		Metadata: pb.EnvironmentMetadata_builder{
			Name: "dev",
			Settings: pb.EnvironmentSettings_builder{
				ImageBuilderVersion: "2025.06",
			}.Build(),
		}.Build(),
	}.Build())

	client, err := NewClientWithOptions(&ClientParams{ControlPlaneClient: mockClient})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer client.Close()

	env, err := client.fetchEnvironment(ctx, "")
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(env.Settings.ImageBuilderVersion).Should(gomega.Equal("2024.10"))
	g.Expect(env.Settings.WebhookSuffix).Should(gomega.Equal("modal.run"))
	g.Expect(mockClient.getCallCount()).Should(gomega.Equal(1))

	env2, err := client.fetchEnvironment(ctx, "")
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(env2).Should(gomega.BeIdenticalTo(env))
	g.Expect(mockClient.getCallCount()).Should(gomega.Equal(1)) // got "" from cache

	envDev, err := client.fetchEnvironment(ctx, "dev")
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(envDev.Settings.ImageBuilderVersion).Should(gomega.Equal("2025.06"))
	g.Expect(mockClient.getCallCount()).Should(gomega.Equal(2)) // got "dev" from server
}

func TestImageBuilderVersion_LocalConfigHasPrecedence(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := context.Background()

	mockClient := &mockEnvironmentClient{}
	mockClient.envs.Store("", pb.EnvironmentGetOrCreateResponse_builder{
		EnvironmentId: "en-test",
		Metadata: pb.EnvironmentMetadata_builder{
			Settings: pb.EnvironmentSettings_builder{
				ImageBuilderVersion: "2024.10",
			}.Build(),
		}.Build(),
	}.Build())

	client, err := NewClientWithOptions(&ClientParams{
		ControlPlaneClient: mockClient,
		Config: &config{
			"default": rawProfile{
				ImageBuilderVersion: "2024.04",
				Active:              true,
			},
		},
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer client.Close()

	version, err := client.imageBuilderVersion(ctx, "")
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(version).Should(gomega.Equal("2024.04"))
	g.Expect(mockClient.getCallCount()).Should(gomega.Equal(0))
}
